internal/oci/runtime_oci.go

```diff
@@ -1220,3 +1231,248 @@ func (r *Runtime) SpoofOOM(c *Container) {
 func ConmonPath(r *Runtime) string {
 	return r.config.Conmon
 }
+
+// CheckpointContainer checkpoints a container.
+func (r *runtimeOCI) CheckpointContainer(c *Container, specgen *rspec.Spec, leaveRunning bool) error {
+	c.opLock.Lock()
```

Comment:
>Here is a potential deadlock. In function `RestoreContainer()`, it will do `c.opLock.Lock()` then call `CreateContainer()` at line 1449 to restore container. If restore failed(like criu returns errors), it will call `DeleteContainer()` at line 200 to delete the container and it will try to do `c.opLock.Lock()` again , hence causes a deadlock.

Sorry for the noisy, switch to a proper account and forward the comment again.

Reply:
Thanks. I fixed that.